### ISSUE TITLE
Failed building wheel for PyICU

### ISSUE THREAD

**Issue by yonilevine** [ovalhub/pyicu#81]
Hi,
Can anyone please help me with this error message. I've been working on it for hours to no avail!! 

Yonis-MacBook-Pro:~ yonilevine$ sudo -H pip install --user git+https://github.com/ovalhub/pyicu
Collecting git+https://github.com/ovalhub/pyicu
  Cloning https://github.com/ovalhub/pyicu to /private/tmp/pip-req-build-5xo_g0m0
Building wheels for collected packages: PyICU
  Running setup.py bdist_wheel for PyICU ... error
  Complete output from command /anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-req-build-5xo_g0m0/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /private/tmp/pip-wheel-eudapbmr --python-tag cp36:
  
  Building PyICU 2.0.3 for ICU 58.2
  
  Adding CXXFLAGS="-I/anaconda3/include" from /anaconda3/bin/icu-config
  Adding LDFLAGS="-Wl,-rpath,/anaconda3/lib -L/anaconda3/lib -licui18n -licuuc -licudata" from /anaconda3/bin/icu-config
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.macosx-10.7-x86_64-3.6
  copying PyICU.py -> build/lib.macosx-10.7-x86_64-3.6
  creating build/lib.macosx-10.7-x86_64-3.6/icu
  copying icu/__init__.py -> build/lib.macosx-10.7-x86_64-3.6/icu
  running build_ext
  building '_icu' extension
  creating build/temp.macosx-10.7-x86_64-3.6
  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/anaconda3/include -arch x86_64 -I/anaconda3/include -arch x86_64 -I/anaconda3/include/python3.6m -c _icu.cpp -o build/temp.macosx-10.7-x86_64-3.6/_icu.o -DPYICU_VER="2.0.3" -I/anaconda3/include
  In file included from _icu.cpp:27:
  ./common.h:38:13: error: unknown type name 'decltype'
      typedef decltype(nullptr) nullptr_t;
              ^
  ./common.h:38:30: error: expected ';' after top level declarator
      typedef decltype(nullptr) nullptr_t;
                               ^
                               ;
  2 errors generated.
  error: command 'gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for PyICU
  Running setup.py clean for PyICU
Failed to build PyICU
Installing collected packages: PyICU
  Running setup.py install for PyICU ... error
    Complete output from command /anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-req-build-5xo_g0m0/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/tmp/pip-record-ptnxlaom/install-record.txt --single-version-externally-managed --compile --user --prefix=:
    
    Building PyICU 2.0.3 for ICU 58.2
    
    Adding CXXFLAGS="-I/anaconda3/include" from /anaconda3/bin/icu-config
    Adding LDFLAGS="-Wl,-rpath,/anaconda3/lib -L/anaconda3/lib -licui18n -licuuc -licudata" from /anaconda3/bin/icu-config
    running install
    running build
    running build_py
    creating build
    creating build/lib.macosx-10.7-x86_64-3.6
    copying PyICU.py -> build/lib.macosx-10.7-x86_64-3.6
    creating build/lib.macosx-10.7-x86_64-3.6/icu
    copying icu/__init__.py -> build/lib.macosx-10.7-x86_64-3.6/icu
    running build_ext
    building '_icu' extension
    creating build/temp.macosx-10.7-x86_64-3.6
    gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/anaconda3/include -arch x86_64 -I/anaconda3/include -arch x86_64 -I/anaconda3/include/python3.6m -c _icu.cpp -o build/temp.macosx-10.7-x86_64-3.6/_icu.o -DPYICU_VER="2.0.3" -I/anaconda3/include
    In file included from _icu.cpp:27:
    ./common.h:38:13: error: unknown type name 'decltype'
        typedef decltype(nullptr) nullptr_t;
                ^
    ./common.h:38:30: error: expected ';' after top level declarator
        typedef decltype(nullptr) nullptr_t;
                                 ^
                                 ;
    2 errors generated.
    error: command 'gcc' failed with exit status 1
    
    ----------------------------------------
Command "/anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-req-build-5xo_g0m0/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /private/tmp/pip-record-ptnxlaom/install-record.txt --single-version-externally-managed --compile --user --prefix=" failed with error code 1 in /private/tmp/pip-req-build-5xo_g0m0/

**Comment by ovalhub** [comment]
On Mon, 11 Jun 2018, yonilevine wrote:

> Hi,
> Can anyone please help me with this error message. I've been working on it for hours to no avail!!
>
>
> Yonis-MacBook-Pro:~ yonilevine$ sudo -H pip install --user git+https://github.com/ovalhub/pyicu
> Collecting git+https://github.com/ovalhub/pyicu
>  Cloning https://github.com/ovalhub/pyicu to /private/tmp/pip-req-build-5xo_g0m0
> Building wheels for collected packages: PyICU
>  Running setup.py bdist_wheel for PyICU ... error
>  Complete output from command /anaconda3/bin/python -u -c "import setuptools, tokenize;__file__='/private/tmp/pip-req-build-5xo_g0m0/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /private/tmp/pip-wheel-eudapbmr --python-tag cp36:
>
>  Building PyICU 2.0.3 for ICU 58.2
>
>  Adding CXXFLAGS="-I/anaconda3/include" from /anaconda3/bin/icu-config
>  Adding LDFLAGS="-Wl,-rpath,/anaconda3/lib -L/anaconda3/lib -licui18n -licuuc -licudata" from /anaconda3/bin/icu-config
>  running bdist_wheel
>  running build
>  running build_py
>  creating build
>  creating build/lib.macosx-10.7-x86_64-3.6
>  copying PyICU.py -> build/lib.macosx-10.7-x86_64-3.6
>  creating build/lib.macosx-10.7-x86_64-3.6/icu
>  copying icu/__init__.py -> build/lib.macosx-10.7-x86_64-3.6/icu
>  running build_ext
>  building '_icu' extension
>  creating build/temp.macosx-10.7-x86_64-3.6
>  gcc -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -I/anaconda3/include -arch x86_64 -I/anaconda3/include -arch x86_64 -I/anaconda3/include/python3.6m -c _icu.cpp -o build/temp.macosx-10.7-x86_64-3.6/_icu.o -DPYICU_VER="2.0.3" -I/anaconda3/include
>  In file included from _icu.cpp:27:
>  ./common.h:38:13: error: unknown type name 'decltype'
>      typedef decltype(nullptr) nullptr_t;
>              ^
>  ./common.h:38:30: error: expected ';' after top level declarator
>      typedef decltype(nullptr) nullptr_t;

This code is compiled when the version of gcc you're using < 4.8.
But it looks like your version is gcc is so old, it doesn't understand
   decltype(), a 'modern' C++ feature.
See https://www.google.com/search?q=gcc+decltype+support
for more information and get a version of gcc > 4.8 or one that support 
decltype().

Andi..

**Comment by johb** [comment]
Hi,

I also ran into this issue with and macOS 10.13.5 and PyICU.

I solved it by installing the current version of gcc from homebrew with
`brew install gcc`
which is
`% /usr/local/bin/gcc-8 --version
gcc-8 (Homebrew GCC 8.1.0) 8.1.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.`

Then using
`env CC=/usr/local/bin/gcc-8 pip install PyICU`

worked for me to successfully install PyICU.
